src/field_theory/galois_field.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyleft. No rights reserved.
+Authors: Johan Commelin
+-/
+
+import field_theory.finite.basic
+import field_theory.separable
+import linear_algebra.finite_dimensional
+
+/-!
+# Galois fields
+
+If `p` is a prime number, and `n` a natural number,
+then `galois_field p n` is defined as the splitting field of `X^(p^n) - X` over `zmod p`.
+It is a finite field with `p ^ n` elements.
+
+## Main definition
+
+* `galois_field p n` is a field with `p ^ n` elements
+
+-/
+
+noncomputable theory
+
+-- move this
+section
+open function
+variables {K L : Type*} [field K] [field L]
+
+variables (K L) [algebra K L]
+
+lemma algebra.char_p_iff (p : ℕ) :
+  char_p K p ↔ char_p L p :=
+(algebra_map K L).char_p_iff_char_p p
+
+end
+
+open polynomial
+
+lemma galois_poly_separable {K : Type*} [field K] (p q : ℕ) [char_p K p] (h : p ∣ q) :
+  separable (X ^ q - X : polynomial K) :=
+begin
+  use [1, (X ^ q - X - 1)],
+  rw [← char_p.cast_eq_zero_iff (polynomial K) p] at h,
+  rw [derivative_sub, derivative_pow, derivative_X, h],
+  ring,
+end
+
+section splitting_field_facts
+
+variables {K : Type*} {L : Type*} [field K] [field L] {f : polynomial K}
+
+lemma exists_finset_of_splits (i : K →+* L) (sep : separable f) : splits i f →
+  ∃ (s : finset L), f.map i = C (i f.leading_coeff) *
+  (s.prod (λ a : L, (X : polynomial L) - C a)) :=
+begin
+  classical,
+  intro sp, cases exists_multiset_of_splits i sp with s h, use s.to_finset, rw h,
+  rw finset.prod_eq_multiset_prod, rw ← multiset.to_finset_eq,
+  apply nodup_of_separable_prod,
+  apply separable.of_mul_right,
+  rw ←h,
+  exact sep.map,
+end
+
+open_locale big_operators
+variables {α : Type*} (s : finset α) (g : α → polynomial K)
+
+-- I have a proof in another project, hope to PR soon - Aaron Anderson
+lemma nat_degree_prod_eq' (h : ∏ i in s, (g i).leading_coeff ≠ 0) :
+  (s.prod g).nat_degree = ∑ i in s, (g i).nat_degree :=
```

Comment:
@awainverse did this land anywhere I didn't find?

Reply:
It was `polynomial.nat_degree_prod'`, but I think `polynomial.nat_degree_prod` works better